### PR TITLE
New checksum.

### PR DESCRIPTION
**Pull request by MikeStall**
Better variance and avoids a bunch of collisions.

### CODE REVIEW

File: src/PAModel/Checksum/IHashMaker.cs [hunk 1]
```diff
@@ -39,57 +38,77 @@ class Sha256HashMaker : IHashMaker, IDisposable
...
 
+        private readonly static byte[] _startObj = new byte[] { (byte)'{' };
+        private readonly static byte[] _endObj = new byte[] { (byte) '}' };
+        private readonly static byte[] _startArray = new byte[] { (byte) '[' };
+        private readonly static byte[] _endArray = new byte[] { (byte) ']' };
+        private readonly static byte[] _null = new byte[] { 254 };
+        private readonly static byte[] _true = new byte[] { 1 };
+        private readonly static byte[] _false  = new byte[] { 0 };
+
+        private readonly static byte[] _marker = new byte[] { 255 };
```
Comment (MikeStall) [inline src/PAModel/Checksum/IHashMaker.cs:49]: This is the magic - we're making the checksum much more rigid now.

File: src/PAModelTests/ChecksumTests.cs [hunk 2]
```diff
@@ -57,44 +58,54 @@ public void ChecksumsUnique()
...
+        const string C20 = "C6_2YUFpLVLEYtdFvV9iLN8F6TM+cWczemMx4m0VEIpfrg=";
 
 
         [DataTestMethod]
         [DataRow(C1, @"' ab\r\ncd'")] // whitespace
         [DataRow(C1, @"'ab\ncd'")] // same
         [DataRow(C2, @"'ab\ncd  '")] // Trailing whitespace is not trimmed. 
         [DataRow(C3, "{'InvariantScript' : 'abc'}")]
-        [DataRow(C3, "[ 'abc']")] // BAD: InvariantScript is skipped
+        [DataRow(C12, "[ 'abc']")] // InvariantScript is skipped
```
Comment (MikeStall) [inline src/PAModelTests/ChecksumTests.cs:90]: The thing to notice here is that we remove a bunch of collisions now (values used to be the same, and now they're different).

File: src/PAModel/Entropy.cs [hunk 3]
```diff
@@ -56,6 +56,11 @@ internal class PropertyEntropy
 
         public PropertyEntropy VolatileProperties { get; set; }
 
+        // Sometimes, an empty LocalDatabaseReferences is serialized as "" or { }.
+        // Need to track which so we can round-trip.
+        // if true: "".   Else,  "{}". 
+        public bool LocalDatabaseReferencesAsEmpty { get; set; }
```
Comment (lesaltzm) [inline src/PAModel/Entropy.cs:62]: Which is the more up-to-date style of serializing this? Is there something we should add to #21 to track this?

Comment (lesaltzm) [inline src/PAModel/Entropy.cs:62]: Also, should this be nullable, for when the LocalDatabaseReference isn't empty at all?

**Review by lesaltzm — APPROVED** [review]

**Pull request merged**